plugins/synaptics-rmi/fu-synaptics-rmi-device.c

```diff
@@ -0,0 +1,1070 @@
+	if (g_getenv ("FWUPD_SYNAPTICS_RMI_VERBOSE") != NULL) {
+		for (guint i = 0; i < priv->functions->len; i++) {
+			FuSynapticsRmiFunction *func = g_ptr_array_index (priv->functions, i);
+			g_debug ("PDT-%02u fn:0x%02x vr:%d sc:%d ms:0x%x "
+				 "db:0x%02x cb:0x%02x cm:0x%02x qb:0x%02x",
+				 i, func->function_number, func->function_version,
+				 func->interrupt_source_count,
+				 func->interrupt_mask,
+				 func->data_base,
+				 func->control_base, func->command_base,
+				 func->query_base);
+		}
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_synaptics_rmi_device_set_mode (FuSynapticsRmiDevice *self,
+				  FuSynapticsRmiHidMode mode,
+				  GError **error)
+{
+	FuSynapticsRmiDevicePrivate *priv = GET_PRIVATE (self);
+	const guint8 data[] = { 0x0f, mode };
+	gint rc;
+
+	fu_common_dump_raw (G_LOG_DOMAIN, "SetMode", data, sizeof(data));
+	rc = ioctl (priv->fd, HIDIOCSFEATURE(sizeof(data)), data);
+	if (rc < 0) {
+		if (rc == -EPERM) {
+			g_set_error_literal (error,
+					     FWUPD_ERROR,
+					     FWUPD_ERROR_PERMISSION_DENIED,
+					     "SetMode: permission denied");
+			return FALSE;
+		}
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "SetMode: ioctl failed with %s", strerror (rc));
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_synaptics_rmi_device_setup (FuDevice *device, GError **error)
+{
+	FuDeviceClass *klass_device = FU_DEVICE_GET_CLASS (device);
+	FuSynapticsRmiDeviceClass *klass_rmi = FU_SYNAPTICS_RMI_DEVICE_GET_CLASS (device);
+	FuSynapticsRmiDevice *self = FU_SYNAPTICS_RMI_DEVICE (device);
+	FuSynapticsRmiDevicePrivate *priv = GET_PRIVATE (self);
+	guint16 addr;
+	guint16 prod_info_addr;
+	guint8 ds4_query_length = 0;
+	guint8 has_build_id_query = FALSE;
+	guint8 has_package_id_query = FALSE;
+	g_autofree gchar *bl_ver = NULL;
+	g_autofree gchar *board_id = NULL;
+	g_autofree gchar *fw_ver = NULL;
+	g_autofree gchar *instance_id = NULL;
+	g_autoptr(GByteArray) f01_basic = NULL;
+	g_autoptr(GByteArray) f01_db = NULL;
+	g_autoptr(GByteArray) f01_product_id = NULL;
+	g_autoptr(GByteArray) f34_ctrl = NULL;
+
+	/* read basic device information */
+	if (!fu_synaptics_rmi_device_set_rma_page (self, 0x00, error))
+		return FALSE;
+
+	/* read PDT */
+	if (!fu_synaptics_rmi_device_scan_pdt (self, error))
+		return FALSE;
+	priv->f01 = fu_synaptics_rmi_device_get_function (self, 0x01, error);
+	if (priv->f01 == NULL)
+		return FALSE;
+	addr = priv->f01->query_base;
+	f01_basic = fu_synaptics_rmi_device_read (self, addr, RMI_DEVICE_F01_BASIC_QUERY_LEN, error);
+	if (f01_basic == NULL) {
+		g_prefix_error (error, "failed to read the basic query: ");
+		return FALSE;
+	}
+	priv->manufacturer_id = f01_basic->data[0];
+	priv->has_lts = (f01_basic->data[1] & RMI_DEVICE_F01_QRY1_HAS_LTS) > 0;
+	priv->has_sensor_id = (f01_basic->data[1] & RMI_DEVICE_F01_QRY1_HAS_SENSOR_ID) > 0;
+	priv->has_query42 = (f01_basic->data[1] & RMI_DEVICE_F01_QRY1_HAS_PROPS_2) > 0;
```

Comment:
I know they're explicit, but are the greater than zeros actually needed here?  given the data type is boolean shouldn't setting it to anything bigger just set it to TRUE?

Reply:
I think a gboolen is actually an int. Is a gboolean value of 0x2 true? Equal to TRUE?

Reply:
Well TRUE resolves to `!FALSE`, so I think that it would work.  Up to you though which seems more correct.